File: cli/src/context_os_events/daemon/state.py
"""Daemon state persistence module.

Tracks daemon state across restarts:
- Start time
- Last sync times
- Event counts
"""

import json
from dataclasses import dataclass, field, asdict
from datetime import datetime
from pathlib import Path
from typing import Optional


@dataclass
class DaemonState:
    """Persisted state across daemon restarts."""

    started_at: datetime = field(default_factory=datetime.now)
    last_git_sync: Optional[datetime] = None
    last_session_parse: Optional[datetime] = None
    file_events_captured: int = 0
    git_commits_synced: int = 0
    sessions_parsed: int = 0

    def to_dict(self) -> dict:
        """Convert state to JSON-serializable dict."""
        return {
            "started_at": self.started_at.isoformat() if self.started_at else None,
            "last_git_sync": self.last_git_sync.isoformat() if self.last_git_sync else None,
            "last_session_parse": self.last_session_parse.isoformat() if self.last_session_parse else None,
            "file_events_captured": self.file_events_captured,
            "git_commits_synced": self.git_commits_synced,
            "sessions_parsed": self.sessions_parsed,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "DaemonState":
        """Create state from dict (loaded from JSON)."""
        def parse_datetime(value: Optional[str]) -> Optional[datetime]:
            if value is None:
                return None
            return datetime.fromisoformat(value)

        return cls(
            started_at=parse_datetime(data.get("started_at")) or datetime.now(),
            last_git_sync=parse_datetime(data.get("last_git_sync")),
            last_session_parse=parse_datetime(data.get("last_session_parse")),
            file_events_captured=data.get("file_events_captured", 0),
            git_commits_synced=data.get("git_commits_synced", 0),
            sessions_parsed=data.get("sessions_parsed", 0),
        )

    def save(self, path: Path) -> None:
        """Save state to JSON file."""
        path.parent.mkdir(parents=True, exist_ok=True)
        with open(path, "w") as f:
            json.dump(self.to_dict(), f, indent=2)

    @classmethod
    def load(cls, path: Path) -> "DaemonState":
        """Load state from JSON file, or return fresh state if not found."""
        if not path.exists():
            return cls()

        try:
            with open(path) as f:
                data = json.load(f)
            return cls.from_dict(data)
        except (json.JSONDecodeError, KeyError):
            # Corrupted state file, start fresh
            return cls()

File: core/src/daemon/state.rs
//! Daemon state persistence module.
//!
//! Tracks daemon state across restarts:
//! - Start time
//! - Last sync times
//! - Event counts

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::path::Path;

/// Daemon state persisted to JSON file.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct DaemonState {
    pub started_at: Option<DateTime<Utc>>,
    pub last_git_sync: Option<DateTime<Utc>>,
    pub last_session_parse: Option<DateTime<Utc>>,
    pub last_chain_build: Option<DateTime<Utc>>,
    #[serde(default)]
    pub file_events_captured: i64,
    #[serde(default)]
    pub git_commits_synced: i64,
    #[serde(default)]
    pub sessions_parsed: i64,
    #[serde(default)]
    pub chains_built: i64,
}

impl DaemonState {
    /// Save state to JSON file.
    pub fn save(&self, path: &Path) -> Result<(), String> {
        // Ensure parent directory exists
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)
                .map_err(|e| format!("Failed to create state directory: {}", e))?;
        }

        let json = serde_json::to_string_pretty(self)
            .map_err(|e| format!("Failed to serialize state: {}", e))?;

        std::fs::write(path, json).map_err(|e| format!("Failed to write state file: {}", e))?;

        Ok(())
    }

    /// Load state from JSON file, or return fresh state if not found.
    pub fn load(path: &Path) -> Result<Self, String> {
        if !path.exists() {
            return Ok(Self::default());
        }

        let content = std::fs::read_to_string(path)
            .map_err(|e| format!("Failed to read state file: {}", e))?;

        // Handle empty or corrupted files
        if content.trim().is_empty() {
            return Ok(Self::default());
        }

        serde_json::from_str(&content).map_err(|e| {
            // Return default on parse error (corrupted file)
            log::warn!("Failed to parse state file, using default: {}", e);
            format!("Failed to parse state file: {}", e)
        })
    }

    /// Load state from JSON file, returning default on any error.
    /// This is the safer version that never fails.
    pub fn load_or_default(path: &Path) -> Self {
        Self::load(path).unwrap_or_default()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    // ========================================================================
    // TDD Cycle 2: DaemonState (4 tests)
    // ========================================================================

    #[test]
    fn test_default_state_has_zero_counts() {
        let state = DaemonState::default();

        assert!(state.started_at.is_none());
        assert!(state.last_git_sync.is_none());
        assert_eq!(state.git_commits_synced, 0);
        assert_eq!(state.sessions_parsed, 0);
        assert_eq!(state.chains_built, 0);
    }

    #[test]
    fn test_state_save_creates_json_file() {
        let temp_dir = TempDir::new().unwrap();
        let state_path = temp_dir.path().join("daemon.state.json");

        let mut state = DaemonState::default();
        state.started_at = Some(Utc::now());
        state.git_commits_synced = 42;

        state.save(&state_path).unwrap();

        assert!(state_path.exists());
        let content = std::fs::read_to_string(&state_path).unwrap();
        assert!(content.contains("42"));
    }

    #[test]
    fn test_state_load_restores_from_json() {
        let temp_dir = TempDir::new().unwrap();
        let state_path = temp_dir.path().join("daemon.state.json");

        // Save state
        let mut original = DaemonState::default();
        original.sessions_parsed = 100;
        original.chains_built = 10;
        original.save(&state_path).unwrap();

        // Load state
        let loaded = DaemonState::load(&state_path).unwrap();

        assert_eq!(loaded.sessions_parsed, 100);
        assert_eq!(loaded.chains_built, 10);
    }

    #[test]
    fn test_state_load_returns_default_if_file_missing() {
        let temp_dir = TempDir::new().unwrap();
        let state_path = temp_dir.path().join("nonexistent.json");

        let state = DaemonState::load(&state_path).unwrap();

        assert_eq!(state.git_commits_synced, 0);
        assert_eq!(state.sessions_parsed, 0);
    }
}
